### ISSUE TITLE
训练结果问题

### ISSUE THREAD

**Issue by asdffgjkalldbd** [zhyingS/TraF-Align#4]
checkpoint里面你们的指标是vehicle:ap@0.5: 0.7332190155898975，ap@0.7: 0.4440083148136422（我用你们的pth文件在测试集上也能跑出类似的数据），
而我train以后跑出来的指标是vehicle  {'ap@0.5': 61.63999999999999, 'ap@0.7': 46.04}，
用了和checkpoint一样的yaml文件v2v4real_Trafalign.yaml，
yaml文件中只修改了一项内容将cls_map: [0,0,0,0,0,1,0,0,0]改成cls_map: [0,0,0,0,0,-1,-1,-1,-1]，即只检测一类，
训练集将v2v4real中所有的train全部下载下来了，
验证集，只下载了一个1g的val
测试集，也仅使用一个1g左右的test
训练则是单卡（3090）训练。

不知道为什么train以后的pth文件，效果差了很多，百思不得其解，是你们对v2v4real_Trafalign.yaml有什么特殊处理吗？

**Comment by zhyingS** [comment]
你好，有两个问题：
1. cls_map应该和我给的保持一致，你的修改方式会改变数据集里的类别和目标数量分布，导致计算出来的AP不一样；
2. 需要用完整的test集进行测试；
我们没有做什么特殊处理，具体关于类别的过滤方式请看代码。

你试试用完全一样的设定重新训练模型，看看是否有差异？

**Comment by asdffgjkalldbd** [comment]
ok，我试试

**Comment by asdffgjkalldbd** [comment]
你们yaml中的设置：cls: ['Car', 'Truck', 'Van', 'ConcreteTruck', 'Bus', 'Pedestrian','ScooterRider', 'BicycleRider', 'Scooter']
  cls_id: [0,1,2,3,4,5,6,7,8]
  cls_map: [0,0,0,0,0,1,0,0,0]
  cls_group: [["Car"]] # [["Car"],["Cyclist","Pedestrian"]]————
  你们给出的checkpoint中的设置：cls:
  - Car
  - Truck
  - Van
  - ConcreteTruck
  - Bus
  - Pedestrian
  - ScooterRider
  - BicycleRider
  - Scooter
  cls_group:
  - - Car
  cls_id:
  - 0
  - 1
  - 2
  - 3
  - 4
  - 5
  - 6
  - 7
  - 8
  cls_map:
  - 0
  - 0
  - 0
  - 0
  - 0
  - 1
  - 0
  - 0
  - 0——————按照yaml文件中原始设置（我对照了一下checkpoint，和yaml文件设置是一致的），会导致模型设定中有两类，0和1（cls_map）；
 但实际上只给出了一类，car（cls_group）；
我之前尝试用yaml原始设置，这会导致模型完全无法训练，ap几乎都为0.

你们的模型不应该这样子设置yaml文件吗，cls_group: [["Cyclist","Pedestrian"]]
为什么checkpoint文件中的  cls_group:- - Car

**Comment by zhyingS** [comment]
你好，

cls_map是用来将原始的类别映射为两类，一类是Vehicle，一类是Cyclist和Pedestrain，分别用0和1来辨识；
cls_group是早期开发center head的时候用的，目前的版本中这个参数只用来计算类别数量；
eval_cls目前只支持vehicle类别，也就是cls_map中为0的类别。如果你想要评测cyclist和pedestrain，请结合代码详细看一下如何适配；

为了复现，除了必要的修改（比如文件路径），请保持所有设定都和原仓库一样，包括yaml、conda环境、数据集。另外学习率要根据你的显卡数量进行缩放，如果还是训练不出来的话，请把训练的曲线放上来我们一起看一下。

**Comment by asdffgjkalldbd** [comment]
我给的压缩包中包含原始设置，修改设置，以及原始设置使用的yaml文件

其中原始设置，代表除了batchsize以外，其余的设置均为改变的设置，但是训练结果ap50以及ap75均为0

而修改设置，代表除了batchsize以外，仅改动cls_map:0，0，0，0，0，-1，-1，-1，-1的设置，训练结果正常一些，vehicle: ap@0.5: 0.5691455006868203 ； ap@0.7: 0.2326993886411285，但还是和checkpoint差距较大。
使用checkpoint中的pth文件在同样的测试集下测试结果为，vehicle: ap@0.5: 0.7332190155898975 ； ap@0.7: 0.4440083148136422

以上训练集（8个压缩包），测试集（3个压缩包），验证集（1个压缩包）均和v2v4real官方发布的一致

我注意到，在我使用修改设置的时候，At epoch 0, the training loss is 5.653090 ；At epoch 0, the validation loss is 2.668187
At epoch 59, the training loss is 3.987568 ；At epoch 58, the validation loss is 1.249526

而在checkpoint中的2024-12-14 06:43:55,025 At epoch 0, the validation loss is 1.784710
2024-12-14 06:43:55,316 At epoch 0, the validation loss is 1.852559
2024-12-14 06:43:55,381 At epoch 0, the validation loss is 1.844264
2024-12-14 06:43:55,821 At epoch 0, the validation loss is 1.820964
2024-12-14 06:43:55,829 At epoch 0, the validation loss is 1.793425
2024-12-14 06:43:56,281 At epoch 0, the validation loss is 1.792036
2024-12-14 06:43:56,805 At epoch 0, the validation loss is 1.771274

2024-12-14 11:05:01,403 At epoch 58, the validation loss is 0.806352
2024-12-14 11:05:01,835 At epoch 58, the validation loss is 0.759504
2024-12-14 11:05:01,939 At epoch 58, the validation loss is 0.762532
2024-12-14 11:05:03,377 At epoch 58, the validation loss is 0.754998
2024-12-14 11:05:03,708 At epoch 58, the validation loss is 0.780639
2024-12-14 11:05:04,010 At epoch 58, the validation loss is 0.827719
2024-12-14 11:05:04,456 At epoch 58, the validation loss is 0.759910

初始和训练结束的验证集损失都小很多，搞不明白为什么


[v2v4real_Trafalign.zip](https://github.com/user-attachments/files/24253500/v2v4real_Trafalign.zip)

**Comment by zhyingS** [comment]
你给的压缩包中，原始设置和修改设置里的两个config.yaml里，训练和验证的文件夹不一样，分别是
原始设置：
- 训练集：trainselect
- 验证集：val

修改设置：
- 训练集：train
- 验证集：test

导致原始设置和我的设定不一样，从log看，validation loss很小，甚至小于1e-5，说明数据集的文件夹设定是有问题的。

然后，修改设置之所以差的多，是因为你设置的cls_map会把=-1的类别过滤掉，目标的数量和分布都变了。

你先保证数据集的设定和我的仓库一致尝试训练。另外你用的gpu数量是1，如果训练不收敛的话，减小学习率试试。

**Comment by asdffgjkalldbd** [comment]
在原始代码文件中，v2v4real_Trafalign.yaml是这样设置数据集的
root_dir: '/data1/dataset_zhiying/V2V4Real/train'
validate_dir: '/data1/dataset_zhiying/V2V4Real/test'
在checkpoint，中的config.yaml文件是这样设置数据集的
name: v2v4real_Trafalign
root_dir: /data1/dataset_zhiying/V2V4Real/train
train_params:
  epoches: 60
  eval_freq: 2
  gpus: 7
  max_cav: 2
  save_freq: 2
  train_batch_size: 2
  val_batch_size: 2
validate_dir: /data1/dataset_zhiying/V2V4Real/test
这样不会导致数据集泄露吗，不是很明白为什么这样设置

**Comment by zhyingS** [comment]
没有明白你的意思。数据集泄漏是指训练集包含测试数据，但这里显然是用的train来训练，test来推理。或许你是指validate_dir应该设置为validation集，这也是完全可以的，不会影响训练结果。

**Comment by asdffgjkalldbd** [comment]
okok了解了
